source/common/http/http1/codec_impl.cc

```diff
@@ -647,72 +610,102 @@ void ServerConnectionImpl::handlePath(HeaderMapImpl& headers, unsigned int metho
   // to disconnect the connection but we shouldn't fire any more events since it doesn't make
   // sense.
   if (active_request_) {
-    const char* method_string = http_method_str(static_cast<http_method>(parser_.method));
+    const char* method_string = parser_->methodName();
 
     // Inform the response encoder about any HEAD method, so it can set content
     // length and transfer encoding headers correctly.
-    active_request_->response_encoder_.isResponseToHeadRequest(parser_.method == HTTP_HEAD);
+    active_request_->response_encoder_.isResponseToHeadRequest(parser_->method() ==
+                                                               static_cast<int>(Method::Head));
 
-    // Currently, CONNECT is not supported, however; http_parser_parse_url needs to know about
+    // Currently, CONNECT is not supported, however; llhttp_parse_url needs to know about
     // CONNECT
-    handlePath(*headers, parser_.method);
+    handlePath(*current_header_map_, parser_->method());
     ASSERT(active_request_->request_url_.empty());
 
-    headers->setMethod(method_string);
+    current_header_map_->setMethod(method_string);
 
     // Determine here whether we have a body or not. This uses the new RFC semantics where the
     // presence of content-length or chunked transfer-encoding indicates a body vs. a particular
     // method. If there is no body, we defer raising decodeHeaders() until the parser is flushed
     // with message complete. This allows upper layers to behave like HTTP/2 and prevents a proxy
     // scenario where the higher layers stream through and implicitly switch to chunked transfer
     // encoding because end stream with zero body length has not yet been indicated.
-    if (parser_.flags & F_CHUNKED ||
-        (parser_.content_length > 0 && parser_.content_length != ULLONG_MAX) || handling_upgrade_) {
-      active_request_->request_decoder_->decodeHeaders(std::move(headers), false);
+    if (parser_->flags() & static_cast<int>(Flags::Chunked) ||
+        (parser_->contentLength() > 0 && parser_->contentLength() != ULLONG_MAX) ||
+        handling_upgrade_) {
+      active_request_->request_decoder_->decodeHeaders(std::move(current_header_map_), false);
 
       // If the connection has been closed (or is closing) after decoding headers, pause the parser
       // so we return control to the caller.
       if (connection_.state() != Network::Connection::State::Open) {
-        http_parser_pause(&parser_, 1);
+        rc = parser_->pause();
       }
 
     } else {
-      deferred_end_stream_headers_ = std::move(headers);
+      deferred_end_stream_headers_ = std::move(current_header_map_);
     }
   }
 
-  return 0;
+  // TODO(dereka) share below with ClientConnectionImpl
+  current_header_map_.reset();
+  header_parsing_state_ = HeaderParsingState::Done;
+
+  // Returning 2 informs llhttp to not expect a body or further data on this connection.
```

Comment:
Is this code common to both http_parser and llhttp? What would http_parser do with 2?